distributed/scheduler.py

```diff
@@ -4082,6 +4084,105 @@ def _to_dict(
+    async def get_cluster_state(
+        self,
+        exclude: Collection[str],
+    ) -> dict:
+        "Produce the state dict used in a cluster state dump"
+        # Kick off state-dumping on workers before we block the event loop in `self._to_dict`.
+        workers_future = asyncio.gather(
+            self.broadcast(
+                msg={"op": "dump_state", "exclude": exclude},
+                on_error="return",
+            ),
+            self.broadcast(
+                msg={"op": "versions"},
+                on_error="ignore",
+            ),
+        )
+        try:
+            scheduler_state = self._to_dict(exclude=exclude)
+
+            worker_states, worker_versions = await workers_future
+        finally:
+            # Ensure the tasks aren't left running if anything fails.
+            # Someday (py3.11), use a trio-style TaskGroup for this.
+            workers_future.cancel()
+
+        # Convert any RPC errors to strings
+        worker_states = {
+            k: repr(v) if isinstance(v, Exception) else v
+            for k, v in worker_states.items()
+        }
+
+        return {
+            "scheduler": scheduler_state,
+            "workers": worker_states,
+            "versions": {"scheduler": self.versions(), "workers": worker_versions},
+        }
+
+    async def dump_cluster_state_to_url(
+        self,
+        url: str,
+        exclude: Collection[str],
+        format: Literal["msgpack", "yaml"],
+        storage_options: Optional[Dict[str, Any]] = None,
+    ) -> None:
+        "Write a cluster state dump to an fsspec-compatible URL. Requires fsspec."
+        try:
+            import fsspec
+        except ImportError:
+            raise ImportError(
+                "fsspec must be installed to dump cluster state directly to a URL"
+            )
+
+        if storage_options is None:
+            storage_options = {}
+
+        if format == "msgpack":
+            import msgpack
+
+            # NOTE: `compression="infer"` will automatically use gzip via the `.gz` suffix
+            mode = "wb"
+            suffix = ".msgpack.gz"
+            if not url.endswith(suffix):
+                url += suffix
+            writer = msgpack.pack
+        elif format == "yaml":
+            import yaml
+
+            mode = "w"
+            suffix = ".yaml"
+            if not url.endswith(suffix):
+                url += suffix
+
+            def writer(state: dict, f):
+                # YAML adds unnecessary `!!python/tuple` tags; convert tuples to lists to avoid them.
+                def tuple_to_list(node):
+                    if isinstance(node, (list, tuple)):
+                        return [tuple_to_list(el) for el in node]
+                    elif isinstance(node, dict):
+                        return {k: tuple_to_list(v) for k, v in node.items()}
+                    else:
+                        return node
+
+                state = tuple_to_list(state)
+                yaml.dump(state, f)
```

Comment:
Yeah I don't like the duplication either. There are a couple subtle differences right now:
* Client runs `tuple_to_list` for both msgpack and yaml because of serialization; `dump_cluster_state_to_url` only runs it for yaml. I'd guess doing it for msgpack isn't necessary anywhere, so this shouldn't matter.
* `dump_cluster_state_to_url` uses fsspec to open the file; client uses plain `open`. This gives you a way to get a cluster dump if you don't want to install fsspec.

Neither of those seem like huge deals to me, so I'll make it a shared function. Just not sure where to put it besides the dreaded `utils.py`.

Reply:
How about a new module? The stuff in https://github.com/dask/distributed/issues/5873 needs to live somewhere as well

>  I'd guess doing it for msgpack isn't necessary anywhere, so this shouldn't matter.

symmetry. Having everything identical regardless of the output format is nice since every code you write works on both files. One thing I actually do often is to convert a received msgpack file to yaml which allows me to grep stuff 🙈

Reply:
> Converting tuples to lists in the input to msgpack doesn't matter though, since the msgpack itself effectively turns tuples into lists in the dumping process (since tuples are represented as lists)

afaik, you'll need to toggle this specifically. IIRC, msgpack load will reconstruct lists (which is horribly slow)